README.md
<p align="center">
  <img src="http://svrless.org/img/svrless-logo.svg" width="400">
</p>

# SVRLESS

Stop writing AWS Cloud Formation templates and start writing code!

SVRLESS is a framework built from scratch for Rails developers who want to dip their toe in the serverless world.
Writing configuration isn't everyone's cup of tea so lets fix that.

The 5 pillars of SVRLESS are:
- Remove the need to write Cloud Formation templates
- Be opinionated in how you structure your serverless project
- Feel like home to Ruby & Rails developers
- Get your project up and running on AWS fast
- Be open to everyone who wants to help or contribute code

## Prerequisite

AWS SAM CLI is required to build and deploy the `template.yaml` generated by this framework.

[Click here for instructions on installing SAM CLI](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/serverless-sam-cli-install.html).

## Getting Started

1. Install the SVRLESS gem


        $ gem install 'svrless'

2. Create a new SVRLESS project


        $ svrless new your-project 

3. Change directory to your project


        $ cd your-project

4. Open `config/routes.rb` and uncomment the resources line

    ```ruby
    resources :post, :comment
    ```

5. Build your project


        $ svrless build


    Notice the newly created `app/functions/posts/controller.rb` and SAM ready `template.yaml`


        $ sam build

6. Start the local SAM server 
        

        $ sam local start-api

   Issue a `POST` request to `http://localhost:3000/posts` to see your lambda spring into action.

## License

The gem is available as open source under the terms of the [MIT License](https://opensource.org/licenses/MIT).


CHANGELOG.md
## [Unreleased]

## [0.2.0]- 2021-02-14

- Template generation is cleaner now.
- Replace Event with ResourceGenerator class.
- Moved template formation from ResourceGenerator into various generator classes.
- Added tests to cover these.

## [0.1.1] - 2021-02-14

- Minor improvements to initial set up

## [0.1.0] - 2021-02-11

- Initial release


test/test_route.rb
require 'svrless'

class TestRoute
  include ResourceGenerator

  resources :post, :comment_resource, :user
end

test/test_helper.rb
# frozen_string_literal: true

$LOAD_PATH.unshift File.expand_path("../lib", __dir__)
require "svrless"
require "minitest/autorun"


test/svrless_test.rb
# frozen_string_literal: true

require_relative "test_helper"
require_relative "test_route"

class SvrlessTest < Minitest::Test
  def test_has_a_version_number
    refute_nil ::Event::VERSION
  end

  def test_event_class
    TestRoute.new
    assert(true)
  end
end


lib/svrless/resource_generator.rb
# frozen_string_literal: true

require "svrless"

module ResourceGenerator
  module ClassMethods
    DEFAULT_ACTIONS = %i[create show update destroy].freeze

    def resources(*args)
      generate_svrless_controller
      generate_cloudformation(klasses: args, actions: DEFAULT_ACTIONS)
      generate_function_files(klasses: args, actions: DEFAULT_ACTIONS)
    end

    def generate_cloudformation(klasses:, actions:)
      output = File.open("template.yaml", "w")
      output << CloudFormation::MainGenerator.new(klasses: klasses, actions: actions).representation
      output.close
    end

    def plural_klass_name(klass)
      klass.to_s.pluralize
    end

    def full_klass_name(klass)
      klass.to_s.camelize.pluralize
    end

    def generate_function_files(klasses:, actions:)
      klasses.each do |klass|
        generate_function_file(klass: klass, actions: actions)
      end
    end

    def generate_function_file(klass:, actions:)
      Dir.mkdir("app") unless Dir.exist?("app")
      Dir.mkdir("app/functions") unless Dir.exist?("app/functions")
      unless Dir.exist?("app/functions/#{plural_klass_name(klass)}")
        Dir.mkdir("app/functions/#{plural_klass_name(klass)}")
      end
      FileUtils.touch("app/functions/#{plural_klass_name(klass)}/controller.rb")
      File.open("app/functions/#{plural_klass_name(klass)}/controller.rb", "w") do |f|
        f.write "require 'svrless_controller'"
        f.write "\n"
        f.write "\n"
        f.write "class #{full_klass_name(klass)} < SvrlessController"
        f.write "\n"
        actions.each_with_index do |method, index|
          f.write "  def self.#{method}(event:, context:)\n"
          f.write "    {\n"
          f.write "      statusCode: 200,\n"
          f.write "      body: { hello: 'world' }.to_json\n"
          f.write "    }\n"
          f.write "  end\n"
          f.write "\n" unless actions.size == index + 1
        end
        f.write  "end"
        f.write  "\n"
      end
    end

    def generate_svrless_controller
      Dir.mkdir("app") unless Dir.exist?("app")
      Dir.mkdir("app/shared") unless Dir.exist?("app/shared")
      Dir.mkdir("app/shared/ruby") unless Dir.exist?("app/shared/ruby")
      Dir.mkdir("app/shared/ruby/lib") unless Dir.exist?("app/shared/ruby/lib")
      File.open("app/shared/ruby/lib/svrless_controller.rb", "w") do |f|
        f.write "class SvrlessController\n"
        f.write "  def self.router(event:, context:)\n"
        f.write "    case event['httpMethod']\n"
        f.write "    when 'GET'\n"
        f.write "      show(event: event, context: context)\n"
        f.write "    when 'POST'\n"
        f.write "      create(event: event, context: context)\n"
        f.write "    when 'DELETE'\n"
        f.write "      destroy(event: event, context: context)\n"
        f.write "    else\n"
        f.write "      update(event: event, context: context)\n"
        f.write "    end\n"
        f.write "  end\n"
        f.write "end\n"
      end
    end
  end

  def self.included(klass)
    klass.extend ClassMethods
  end
end


lib/svrless/generators/cloud_formation/main_generator.rb
# frozen_string_literal: true

require "svrless"
require_relative "layer_generator"
require_relative "function_generator"

module CloudFormation
  # The main outer shell of SAM template generators
  class MainGenerator
    AWS_TEMPLATE_VERSION = "2010-09-09"
    AWS_TRANSFORM_VERSION = "AWS::Serverless-2016-10-31"

    attr_accessor :klasses, :actions

    def initialize(klasses:, actions:)
      @klasses = klasses
      @actions = actions
    end

    def representation
      main_template_hash.deep_transform_keys { |k| k.to_s.capitalize.camelize }.to_yaml
    end

    def hash_representation
      main_template_hash
    end

    private

    def main_template_hash
      {
        a_w_s_template_format_version: AWS_TEMPLATE_VERSION,
        transform: AWS_TRANSFORM_VERSION,
        resources: resources_hash
      }
    end

    def resources_hash
      resources = {}
      resources.merge! function_hashes
      resources.merge! CloudFormation::LayerGenerator.new.hash_representation
      resources
    end

    def function_hashes
      functions = {}
      @klasses.each do |klass|
        functions.merge! CloudFormation::FunctionGenerator.new(klass: klass, actions: actions).hash_representation
      end
      functions
    end
  end
end


lib/svrless/generators/cloud_formation/event_generator.rb
# frozen_string_literal: true

require "svrless"

module CloudFormation
  # Responsible for generating a hash which will eventually look like this:
  # Create:
  #   Type: Api
  #   Properties:
  #     Method: post
  #     Path: "/posts"
  #
  # Input: klass: Class, action: :show
  # Input: klass: Post, action: :update
  class EventGenerator
    attr_accessor :klass, :action

    def initialize(klass:, action:)
      @klass = klass
      @action = action
    end

    def hash_representation
      {
        "#{@action}": {
          type: "Api",
          properties: {
            method: http_method,
            path: api_path
          }
        }
      }
    end

    def http_method
      case @action
      when :create
        "post"
      when :show
        "get"
      when :update
        "put"
      else
        "delete"
      end
    end

    def api_path
      return "/#{restify_klass_name}" if @action == :create

      "/#{restify_klass_name}/{id}"
    end

    def restify_klass_name
      @klass.to_s.underscore.pluralize.dasherize
    end
  end
end


lib/svrless/generators/cloud_formation/function_generator.rb
# frozen_string_literal: true

require "svrless"

require_relative "event_generator"

module CloudFormation
  # Responsible for generating a hash which will eventually look like this:
  class FunctionGenerator
    attr_accessor :klass, :actions

    def initialize(klass:, actions:)
      @klass = klass
      @actions = actions
    end

    def hash_representation
      {
        "#{plural_klass_name}_controller": {
          type: "AWS::Serverless::Function",
          properties: properties_hash
        }
      }
    end

    private

    def properties_hash
      {
        code_uri: "app/functions/#{plural_klass_name}",
        events: events_hash,
        function_name: "#{plural_klass_name}_controller".capitalize.camelize,
        handler: "controller.#{klass_name}.router",
        Layers: [{ Ref: "SvrlessLayer" }],
        runtime: "ruby2.7"
      }
    end

    def events_hash
      hash = {}
      @actions.each do |action|
        hash.merge! EventGenerator.new(klass: @klass, action: action).hash_representation
      end
      hash
    end

    def plural_klass_name
      @klass.to_s.pluralize.underscore
    end

    def klass_name
      @klass.to_s.camelize.pluralize
    end
  end
end


lib/svrless/event.rb
# frozen_string_literal: true

require "active_support/all"
require "yaml"

module Event
  module ClassMethods
    def events(*args, resource)
      raise StandardError, "Event: You need to supply at least one argument" if args.empty?

      events = args
      generate(events, resource)
    end

    def resource(args)
      resource = args
      events(:create, :show, :update, :destroy, resource)
      generate_function_files(:create, :show, :update, :destroy, resource)
      generate_svrless_controller
    end

    def generate(events, resource)
      puts "Generating resource..."
      # make_function_files
      output = File.open("template.yaml", "w")
      output << main_representation(events, resource)
      output.close
      puts "DONE!"
    end

    def main_representation(events, resource)
      {
        a_w_s_template_format_version: "2010-09-09",
        transform: "AWS::Serverless-2016-10-31",
        resources: resources_hash(events, resource)
      }.deep_transform_keys { |k| k.to_s.capitalize.camelize }.to_yaml
    end

    def resources_hash(events, resource)
      layer.merge(function_hash(events, resource))
    end

    def layer
      {
        svrless_layer: {
          type: "AWS::Serverless::LayerVersion",
          properties: {
            layer_name: "svrless-layer",
            description: "All the business logic should go in here",
            content_uri: "app/shared/.",
            compatible_runtimes: ["ruby2.7"]
          }
        }
      }
    end

    def function_hash(events, resource)
      {
        "#{plural_classname(resource)}_controller": {
          type: "AWS::Serverless::Function",
          properties: {
            code_uri: "app/functions/#{plural_classname(resource)}",
            events: hash_of_events(events, resource),
            function_name: "#{plural_classname(resource)}_controller".capitalize.camelize,
            handler: "controller.#{full_classname(resource)}.router",
            Layers: [{ Ref: "SvrlessLayer" }],
            runtime: "ruby2.7"
          }
        }
      }
    end

    def hash_of_events(events, resource)
      h = {}
      events.each do |event|
        h.merge!(
          {
            "#{event}": {
              type: "Api",
              properties: {
                method: method_from_event(event: event),
                path: path_from_event(event: event, resource: resource)
              }
            }
          }
        )
      end
      h
    end

    def plural_classname(resource)
      resource.to_s.pluralize
    end

    def method_from_event(event:)
      case event
      when :create
        "post"
      when :show
        "get"
      when :update
        "put"
      when :destroy
        "delete"
      end
    end

    def path_from_event(event:, resource:)
      case event
      when :create
        "/#{plural_classname(resource)}"
      when :show
        "/#{plural_classname(resource)}/{id}"
      when :update
        "/#{plural_classname(resource)}/{id}"
      when :destroy
        "/#{plural_classname(resource)}/{id}"
      end
    end

    def full_classname(resource)
      "#{plural_classname(resource)}_controller".camelize
    end

    def generate_function_files(*args, resource)
      Dir.mkdir("app") unless Dir.exist?("app")
      Dir.mkdir("app/functions") unless Dir.exist?("app/functions")
      unless Dir.exist?("app/functions/#{plural_classname(resource)}")
        Dir.mkdir("app/functions/#{plural_classname(resource)}")
      end
      FileUtils.touch("app/functions/#{plural_classname(resource)}/controller.rb")
      File.open("app/functions/#{plural_classname(resource)}/controller.rb", "w") do |f|
        f.write "require 'svrless_controller'"
        f.write "\n"
        f.write "\n"
        f.write "class #{full_classname(resource)} < SvrlessController"
        f.write "\n"
        args.each_with_index do |method, index|
          f.write "  def self.#{method}(event:, context:)\n"
          f.write "    {\n"
          f.write "      statusCode: 200,\n"
          f.write "      body: { hello: 'world' }.to_json\n"
          f.write "    }\n"
          f.write "  end\n"
          f.write "\n" unless args.size == index + 1
        end
        f.write  "end"
        f.write  "\n"
      end
    end

    def generate_svrless_controller
      Dir.mkdir("app") unless Dir.exist?("app")
      Dir.mkdir("app/shared") unless Dir.exist?("app/shared")
      Dir.mkdir("app/shared/ruby") unless Dir.exist?("app/shared/ruby")
      Dir.mkdir("app/shared/ruby/lib") unless Dir.exist?("app/shared/ruby/lib")
      File.open("app/shared/ruby/lib/svrless_controller.rb", "w") do |f|
        f.write "class SvrlessController\n"
        f.write "  def self.router(event:, context:)\n"
        f.write "    case event['httpMethod']\n"
        f.write "    when 'GET'\n"
        f.write "      show(event: event, context: context)\n"
        f.write "    when 'POST'\n"
        f.write "      create(event: event, context: context)\n"
        f.write "    when 'DELETE'\n"
        f.write "      destroy(event: event, context: context)\n"
        f.write "    else\n"
        f.write "      update(event: event, context: context)\n"
        f.write "    end\n"
        f.write "  end\n"
        f.write "end\n"
      end
    end
  end

  def self.included(klass)
    klass.extend ClassMethods
  end
end
